# Исследование надежности заемщиков


Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Описание данных
* children — количество детей в семье
* days_employed — общий трудовой стаж в днях
* dob_years — возраст клиента в годах
* education — уровень образования клиента
* education_id — идентификатор уровня образования
* family_status — семейное положение
* family_status_id — идентификатор семейного положения
* gender — пол клиента
* income_type — тип занятости
* debt — имел ли задолженность по возврату кредитов
* total_income — ежемесячный доход
* purpose — цель получения кредита

## Открываю таблицу и изучаю общую информацию о данных

**Импортирую библиотеку pandas. Считываю данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Вывожу первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Вывожу основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Вывожу количество пропущенных значений для каждого столбца. Использую комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце я обработаю на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. Таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обрабатываю значения в этом столбце: заменяю все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Для каждого типа занятости вывожу медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')
#data.groupby('income_type')['total_income'].agg('median')
#data[data['days_employed'].isna()].groupby('income_type').count()

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставляю их как есть. Тем более этот столбец не понадобится для исследования.

**Вывожу перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**В столбце `children` есть два аномальных значения. Удаляю строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Ещё раз вывожу перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Заполняю пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Убеждаюсь, что все пропуски заполнены. Проверяю себя и ещё раз вывожу количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Заменяю вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Обрабатываю неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Привожу их к нижнему регистру. Проверяю остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Вывожу на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удаляю их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**На основании диапазонов, указанных ниже, создаю в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Использую собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Вывожу на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Создаю функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Использую собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучаю данные в столбце `purpose` и определяю, какие подстроки помогут правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследую данные и отвечаю на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
#0 - платят кредит в срок
#1 - не платят кредит в срок
#children_groupby = data.groupby('children')['debt'].mean()
#children_unique = data['children'].unique()
#children_grouped = data.groupby(['children', 'debt']).agg({'children':'sum'})
##print(children_groupby.sort_values())# Ваш код будет здесь. Вы можете создавать новые ячейки.

def data_pivot(df, category_column):
    table_total = df.pivot_table(index = category_column, values = 'debt', aggfunc=['count', 'sum', 'mean'])
    table_total.columns = ['Всего кредитополучателей', 'Всего должников', 'Доля должников']
    table_total = table_total.sort_values(by='Доля должников')
    display(table_total)
data_pivot(data, 'children')

Всего кредитополучателей  Всего должников  Доля должников
children                                                           
5                                9                0        0.000000
0                            14091             1063        0.075438
3                              330               27        0.081818
1                             4808              444        0.092346
2                             2052              194        0.094542
4                               41                4        0.097561

**Вывод:** Судя по данным выше, нет никакой прямой зависимости между возвратом кредита в срок и количеством детей, но кое-какие моменты выделить всё же можно. Люди без детей являются наиболее ответственными заёмщиками, а люди с 4 детьми наоборот, наименее ответственными

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
#0 - платят кредит в срок
#1 - не платят кредит в срок

#family_unique = data['family_status'].unique
#family_groupby = data.groupby('family_status')['debt'].mean()
#print(family_groupby.sort_values())# Ваш код будет здесь. Вы можете создавать новые ячейки.
data_pivot(data, 'family_status')

Всего кредитополучателей  Всего должников  \
family_status                                                      
вдовец / вдова                              951               63   
в разводе                                  1189               84   
женат / замужем                           12261              927   
гражданский брак                           4134              385   
Не женат / не замужем                      2796              273   

                       Доля должников  
family_status                          
вдовец / вдова               0.066246  
в разводе                    0.070648  
женат / замужем              0.075606  
гражданский брак             0.093130  
Не женат / не замужем        0.097639

**Вывод:** Люди, которые были в браке, выплачивают кредит в срок чаще всего, это касается категории 'вдовец/вдова' и 'в разводе', люди из первой категории имеют лучшую статистку по выпалатам кредитов в срок. Люди, состоящие в браке на данный момент, имеют статистику чуть хуже, при этом люди в категории 'женат/замужем' выплачивают кредит в срок чаще, чем люди из категории 'гражданский брак'. И худшая статистика по выплатам кредита в срок у людей, которые никогда не были женаты или замужем

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
#0 - платят кредит в срок
#1 - не платят кредит в срок

#total_income_category = data['total_income_category'].unique()
#total_income_category_groupby = data.groupby('total_income_category')['debt'].mean()
#print(total_income_category_groupby.sort_values())# Ваш код будет здесь. Вы можете создавать новые ячейки.
data_pivot(data, 'total_income_category')

Всего кредитополучателей  Всего должников  \
total_income_category                                              
D                                           349               21   
B                                          5014              354   
A                                            25                2   
C                                         15921             1353   
E                                            22                2   

                       Доля должников  
total_income_category                  
D                            0.060172  
B                            0.070602  
A                            0.080000  
C                            0.084982  
E                            0.090909

**Вывод:** Судя по данным выше, нет никакой прямой зависимости между возвратом кредита в срок и уровнем дохода, но можно выделить клиентов из категории B, так как они являются наиболее платёжеспособными клиентами, и чаще всего возвращают кредит в срок. К категотрии B относятся клиенты с доходом 200001–1000000. Клиентов из категории С рассматривать не стоит, так как они достаточно редко возвращают кредит в срок, к ним относятся клиенты с доходом 50001–200000. Рассмотреть их можно только в совокупности с другими положительными признаками для выплата кредита в срок

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
#0 - платят кредит в срок
#1 - не платят кредит в срок

#purpose_category_gtoupby = data.groupby('purpose_category')['debt'].mean()
#print(purpose_category_gtoupby.sort_values())# Ваш код будет здесь. Вы можете создавать новые ячейки.
data_pivot(data, 'purpose_category')

Всего кредитополучателей  Всего должников  \
purpose_category                                                      
операции с недвижимостью                     10751              780   
проведение свадьбы                            2313              183   
получение образования                         3988              369   
операции с автомобилем                        4279              400   

                          Доля должников  
purpose_category                          
операции с недвижимостью        0.072551  
проведение свадьбы              0.079118  
получение образования           0.092528  
операции с автомобилем          0.093480

**Вывод:** Судя по данным ниже, люди, берущие кредит на опреиации с недвижимостью, выплачивают кредит в срок чаще всего. Возможно, это связано с тем, что они более ответственные. Те люди, которые берут кредит на проведение свадьбы, также достаточно
часто выплачивают кредит в срок, но реже, чем люди в категории 'операции с недвижимостью'. Люди, берущие кредит для получения образования и операций с автомобилем выплачивают кредит в срок реже остальных категорий, поэтому я бы не советовал выдавать им кредит, особенно на операции с автомобилями, статистика их выплаты кредита в срок самая худшая, но так как таких клиентов около 25%, стоит рассмотреть и их, но в совокупности с другими признаками, о которых я напишу в выводе

#### 3.5 Какие возможные причины появления пропусков в исходных данных?

*Ответ:* пропуски присутствуют в столбцах общего трудового стажа в днях и ежемесячного дохода, причём в обоих столбацх пропусков одинаковое количество, а именно 2174. Также, мы знаем из столбца тип занятости все уникальные значения, а именно безработный, в декрете, госслужащий, компаньон, пенсионер, предприниматель, сотрудник и студент. 
Если в коде прописать следующую строчку:
data[data['days_employed'].isna()].groupby('income_type').count()
то можно заметить, что у нас отсутствуют данные ежемесячного дохода и общего трудового стажа по предпринимателям, госслужащим, сотрудникам, компаньонам и пенсионерам, в связи с этим мы имеем такое большое количество пропусков. Помимо того, что я описал выше, также в данных могут присутствовать пропуски из-за человеческого и технического фактора. Люди могли просто сообщать не всю информацию, возможно, где-то они ошибались, возможно, просто эта информация была необходима. Технический фактор также мог сыграть свою роль. Объединение таблиц или другие операции могли удалить часть данных из таблиц.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* совсем удалять недостающие данные будет не лучшим решением, так как мы потеряем 9,9% данных. В случае с медианными значениями мы получаем цифры, которые максимально были бы похожими на реальные показатели. Мы выбираем именно медианные показатели, а не средние, так как в этом случае данные не будут сильно отличаться в случае, если в доступных данных из таблицы будут значения, сильно отличающиеся от общей выборки

### Шаг 4: общий вывод.

Вывод: при решении, одобрять ли кредит клиенту, в первую очередь нужно учитывать его семейное положение, которое непосредственно влияет на погашение кредита в срок. Наилучшими клиентами, которые с большей вероятностью будут погашать кредит в срок, будут люди без детей, состоявшие в браки ранее, с уровнем дохода в 200001–1000000, а также их кредит должен быть для операций с недвижимостью. Самыми худшими же клиентами будут не женатые/замужние с детьми, их уровень дохода будет составлять от 50001 до 200000, а также их цель кредита это соверешние операций с машинами. 